Setup tome-sync transforming

bkjohnson opened the pull request:
## Description

The data we get from tome-sync isn't a 1:1 match with the data we are currently getting from graphql, so we need to transform it a bit in order for it to work with the templates.

There are certain "global" transformations that need to be done, such as converting snake_case to camelCase, but additionally each target_id may need its own transformation.  This attempts to outline that pattern in a similar way to how the target_id-specific whitelisting works.

## Testing done

Eyeballs :grimacing: 

## Screenshots

### Excerpt from `pages.json`

![image](https://user-images.githubusercontent.com/2008881/67792995-56ec9000-fa37-11e9-8c29-c6bbf1ec7e72.png)

This excerpt has 15 total properties

### Output for 1 file

<!--
![image](https://user-images.githubusercontent.com/2008881/67793239-b2b71900-fa37-11e9-8ae4-724216a954a3.png)
-->

![image](https://user-images.githubusercontent.com/2008881/67994375-3290df00-fc02-11e9-8f68-6edffd15c1c6.png)


This transformed output has 13 properties in total, with the following ones missing:
- `entityUrl`
- `entityId`

`entityUrl` is a pretty important piece that we still need to create.  `entityUrl.breadcrumbs` is an array containing parent items, and the child `page` entity doesn't have a reference to its parent.  We need to come up with a strategy for handling this - some options are:
- Have the transformer find and open the parent file recursively
- Put a placeholder in & add this to a list to be revisited later once all files have been opened & processed

The output for several of the properties doesn't match because we don't have filters for any of the other properties.

## Acceptance criteria
- [ ] The templates don't need to be modified
- [ ] Set up a scaffolding so each content model type has its own transformations
- [ ] Transform the properties for the page content model

<!--
## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
-->

bkjohnson committed:
Recursively convert entity to camelCase

bkjohnson committed:
Fix bug in toCamel

Non-snake keys (like `title`) weren't being handled properly

bkjohnson committed:
Collapse the title value in a questionable way

This doesn't throw any errors with the current output but I'm not sure
if we can rely on the same behavior/expectation in the future.

bkjohnson committed:
Move transform code into separate file

As with filters, we may need lots of different transformations for
specific node types

bkjohnson committed:
Begin transformer for page entities

bkjohnson committed:
Collapse fieldIntroText

bkjohnson committed:
Transform some dates

bkjohnson committed:
Transform fieldAlert & fieldDescription

bkjohnson committed:
Transform moderation_state -> entityPublished

bkjohnson committed:
Set entityBundle to 'page'

Its possible that this kind of behavior doesn't need to happen in each
individual transformer - maybe we could set entityBundle to
contentModelType before we call the transformer if all entities have it

bkjohnson committed:
Use more destructuring

bkjohnson committed:
Add helper for collapsing arrays

More verbose, but also more explicit & hopefully less error prone

kelsonic commented on src/site/stages/build/page-builder/transformers.js:
```diff
@@ -0,0 +1,109 @@
+
+// Recursively replaces all snake_case properties in an object
+// with camelCased properties
+function toCamel(obj) {
```
These are really fun functions to write! However, I think we may want to go for [Lodash's `camelCase` method](https://lodash.com/docs/4.17.15#camelCase) here since it does a lot of null checking built-in to prevent runtime errors. What do you think @cvalarida?

cvalarida replied:
I'm typically a fan of using code somebody else wrote. :smile: In this case, though, what really matters is performance, so I wrote up [a little benchmark](http://jsben.ch/UYzYp).

Looks like lodash's is pretty quick. I'm in favor of using lodash for this.
![image](https://user-images.githubusercontent.com/12970166/67803229-d4b99700-fa49-11e9-8e27-34ce3433ce41.png)

bkjohnson replied:
It helps to know what dependencies we are already using :man_facepalming:

bkjohnson replied:
Going off of the docs and my limited testing, it looks like `camelCase` is meant only for strings, so having the benchmarks pass an object to it may be a bit misleading.  This is what I ended up going with:

```js
_.mapKeys(entity, (v, k) => _.camelCase(k));
```

The [benchmark](http://jsben.ch/OECaa) is a bit different, but lodash still seems ok:

![image](https://user-images.githubusercontent.com/2008881/67814902-9d0b1900-fa62-11e9-961a-84002b167f50.png)

kelsonic commented on src/site/stages/build/page-builder/transformers.js:
```diff
@@ -0,0 +1,109 @@
+ * array with one item which is an object with a single `value` property
+ *
+ */
+function arrayCollapse(arr) {
```
If I understand it correctly, I think this function should be named differently to `takeLastValue` or something like that. When I read `arrayCollapse` I had thought that it meant something like [`flattenDeep`](https://lodash.com/docs/4.17.15#flattenDeep) but since it returns the array's last object's `value` property, we may want to find a more fitting name for future devs :)

cvalarida replied:
How do you feel about breaking this out into two functions?
1) Translate `{ value: 'anything' }` -> `'anything'`
2) Transform arrays into the most fitting data structure
```js
function transmogrifyArray(arr) {
  if (arr.length === 0) return null;
  if (arr.length === 1) return arr[0];
  return arr;
}
```

This way, if we have an array of `{ value: 'whatever' }` we'll end up with the array we expect while still not having to mess with `value`.

bkjohnson replied:
I'm not convinced that breaking it out into two functions gets us anything.

@ElijahLynn may be able to weigh in, but it seems like the `[ { value: 'foo' } ]` pattern is a standard Drupal thing.  If that's the case, then what about something like:


```js

function getDrupalValue(arr) {
  if (arr.length === 0) return null;
  if (arr.length === 1) return arr[0].value;
}
```

kelsonic commented on src/site/stages/build/page-builder/transformers.js:
```diff
@@ -0,0 +1,109 @@
+  transformed.entityPublished = published === 'published';
+  delete transformed.moderationState;
+
+  if (Array.isArray(fieldDescription) && fieldDescription.length === 0) {
```
You could use [Lodash's `isEmpty`](https://lodash.com/docs/4.17.15#isEmpty) here if you want. It'd look something like: `if (isEmpty(fieldDescription)) {` 👍

kelsonic commented on src/site/stages/build/page-builder/transformers.js:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  if (
+    Array.isArray(fieldAlert) &&
```
[Same here](https://github.com/department-of-veterans-affairs/vets-website/pull/10973/files#r340257303) :)

bkjohnson replied:
I think this one is a bit different.  `fieldAlert` might be `[[]]` which isn't considered empty.  However, this worked:

```js
if (_.isEmpty(_.flatten(fieldAlert))) {
  // ...
}
```

cvalarida reviewed (commented):
I think you're on the right track. Looks like it still needs to do the transformation _after_ it fills in the references with the sub-entities, but the approach is good so far!

Also, it might be worth breaking out each content model transformer to its own `.js` file and have one for shared helpers. I can see that `transformers.js` growing pretty large pretty quickly.

bkjohnson committed:
Use lodash to convert to camelCase

This doesn't handle deep nesting at the moment

bkjohnson committed:
Use lodash to check for empty arrays

bkjohnson committed:
Temp fix - ensure transformer isn't undefined

bkjohnson committed:
Create a helper for getting the transformer func

I'm not super thrilled with how the helper is implemented, but we want
to log a warning if the object returns undefined for the entity type.
If this is the case, we also need our "transformer" to just return the
entity unmodified instead

bkjohnson commented:
> Looks like it still needs to do the transformation _after_ it fills in the references with the sub-entities

@cvalarida Can you elaborate on this? Due to the recursive `assembleEntityTree` all of the children will get expanded -> filtered -> transformed in a breadth first way, right?

cvalarida commented:
Currently, the reference expansion is done depth-first.

Regardless, I think if we transform the entity before [returning it](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/stages/build/tome-sync-page-builder.js#L59), that'll do the trick. The result of [`assembleEntityTree`](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/stages/build/tome-sync-page-builder.js#L23) will then always be a completely transformed entity—not only the current entity, but also all sub-entities will have all their transformations.

I think that might be what you're saying in your edit, yeah.

Does that clear anything up? Words are hard. :sweat_smile:

bkjohnson commented on src/site/stages/build/page-builder/transformers.js:
```diff
@@ -36,15 +38,11 @@ function pageTransform(entity) {
-    fieldAlert.length === 1 &&
-    fieldAlert[0].length === 0
-  ) {
+  if (_.isEmpty(_.flatten(fieldAlert))) {
```
Unless I'm doing something wrong locally, I don't think this works with node unfortunately.

bkjohnson committed:
Be more specific with lodash imports

bkjohnson committed:
Rearrange things so transformers are in own dir

I'd like to find a better place to put the arrayCollapse helper.  Had
some difficulty with it for some reason

bkjohnson committed:
Refactor helper func for getting drupal value

bkjohnson committed:
Include metatag info & transform

bkjohnson committed:
Merge branch 'master' into tome-sync-transform

bkjohnson committed:
Move files into the right places

bkjohnson committed:
Remove TODO & add newline

cvalarida commented on src/site/stages/build/process-cms-exports/transform.js:
```diff
@@ -0,0 +1,55 @@
+  const entityTransformer = getEntityTransformer(entityType);
+
+  // Convert all snake_case keys to camelCase
+  const transformed = mapKeys(entity, (v, k) => camelCase(k));
```
Ooh, cool. Hadn't seen `mapKeys` before.

cvalarida commented on src/site/stages/build/process-cms-exports/transform.js:
```diff
@@ -0,0 +1,55 @@
+const pageTransform = require('./transformers/page');
+
+const transformers = {
+  page: pageTransform,
```
Alternatively, you can drop the `Transform` and do
```js
const transformers = {
  page,
};
```

cvalarida commented on src/site/stages/build/process-cms-exports/transform.js:
```diff
@@ -0,0 +1,55 @@
+function getEntityTransformer(entityType) {
+  let entityTransformer = entity => entity;
+
+  if (entityType in transformers) {
```
:thinking: Why use `in` instead of `if (transformers[entityType]) { ... }`?

bkjohnson replied:
I don't have a particularly strong opinion on this, but I feel like using `in` makes things a bit more explicit.  `if(transformers[entityType])` could be checking if the property value is `false`, rather than `undefined`.

cvalarida commented on src/site/stages/build/process-cms-exports/transformers/page.js:
```diff
@@ -0,0 +1,59 @@
+}
+
+function pageTransform(entity) {
+  const transformed = entity;
```
If you're trying to not modify `entity`, this won't quite get you there. You'd need `cloneDeep` or something like it.

I think you should keep it like you've got it right now, though, since `cloneDeep` will slow things down a bit.

Just thought I'd mention it is all.

bkjohnson replied:
Yeah, this _does_ look a bit weird.  I know that the variables are essentially the same - I think initially I meant `transformed` to be more of an indicator of "this is changing" rather than any sort of true separation of objects.

cvalarida commented on src/site/stages/build/process-cms-exports/transformers/page.js:
```diff
@@ -0,0 +1,59 @@
+    metatag: { value: metaTags },
+  } = entity;
+  // collapse title
+  // Question: Can we always assume that title is an array of one item, with that item being an object with a `value` key?
```
We can once we get the [JSON schema validation](https://github.com/department-of-veterans-affairs/vets-website/pull/10982) in place. :wink:

bkjohnson commented:
There are a few touch ups that I'd like to make but for now I'll merge this so we can get to work on the other transformation tickets.
